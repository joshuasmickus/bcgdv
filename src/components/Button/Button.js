import React from 'react';
import { func, node, oneOfType, string } from 'prop-types';


export const Button = ({ className, onClick, children }) => (
  <button className={`button ${className}`} onClick={onClick}>{children}</button>
);


Button.propTypes = {
  className: string,
  onClick: func.isRequired,
  children: oneOfType([
    string,
    node
  ]).isRequired
};

Button.defaultProps = {
  className: ''
}
