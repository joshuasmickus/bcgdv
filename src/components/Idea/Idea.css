.idea {
  border: 1px solid black;
  -moz-box-sizing:border-box;
  -webkit-box-sizing:border-box;
  box-sizing: border-box;
  display: inline-block;
  font-size: 1rem;
  height: 150px;
  list-style: none;
  margin: 10px;
  min-width: 150px;
  padding: 8px;
  width: 150px;
}

.delete-button,
.delete-button:focus {
  background: none;
  border: 0;
  float: right;
  outline: 0;
  position: relative;
  visibility: hidden;
}

.idea:hover .delete-button {
  visibility: visible;
}

.delete-button img {
  display: block;
  height: 20px;
  width: 20px;
}
