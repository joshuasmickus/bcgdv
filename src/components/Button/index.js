import './Button.css';

export { Button as default } from './Button';
