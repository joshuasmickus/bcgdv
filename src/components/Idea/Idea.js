import React, { PureComponent } from 'react';
import { func, shape } from 'prop-types';

import { ideaShape } from '../../Constants/PropTypes';
import { deleteIdea, editIdea } from '../../redux/ideas';

import Button from '../Button';

import deleteIcon from './delete.png';

export class Idea extends PureComponent {
  handleDelete = () => {
    const { actions, id } = this.props;

    actions.deleteIdea(id);
  }

  handleDoubleClick = () => {
    const { actions, id } = this.props;

    actions.editIdea(id);
  }

  render() {
    const { id, title, body } = this.props;

    return (
      <li className={`idea ${id}`} onDoubleClick={this.handleDoubleClick}>
        <h3>{title}</h3>
        <p>{body}</p>
        <Button className="delete-button" onClick={this.handleDelete}>
          <img src={deleteIcon} alt="Delete" title="Delete" />
        </Button>
      </li>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  actions: {
    deleteIdea: (id) => dispatch(deleteIdea.request(id)),
    editIdea: (id) => dispatch(editIdea.request(id))
  }
})

Idea.propTypes = {
  ...ideaShape,
  actions: shape({
    deleteIdea: func.isRequired,
    editIdea: func.isRequired
  }).isRequired
};
