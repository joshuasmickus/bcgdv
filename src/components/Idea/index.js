import { connect } from 'react-redux';

import { Idea, mapDispatchToProps } from './Idea';

import './Idea.css';

export default connect(null, mapDispatchToProps)(Idea);
