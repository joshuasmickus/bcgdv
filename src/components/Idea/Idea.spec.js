import React from 'react';
import { shallow } from 'enzyme';

import Button from '../Button';

import { Idea } from './Idea';

const deleteIdeaMock = jest.fn();
const editIdeaMock = jest.fn();
const actions = {
  deleteIdea: deleteIdeaMock,
  editIdea: editIdeaMock
};
const ID = 'some Id';
const requiredProps = {
  id: ID,
  createdDate: 'someDate',
  body: 'the body',
  title: 'this title here',
  actions
};

describe('Given an Idea component', () => {
  describe('when it is rendered', () => {
    beforeEach(() => {
      deleteIdeaMock.mockReset();
      editIdeaMock.mockReset();
    });
    it('should match the snapshot', () => {
      const wrapper = shallow(<Idea {...requiredProps} />);

      expect(wrapper.getElement()).toMatchSnapshot();
    });

    describe('and delete idea is clicked', () => {
      it('should trigger a deleteIdea action', () => {
        const wrapper = shallow(<Idea {...requiredProps} />);

        wrapper.find(Button).simulate('click');

        expect(deleteIdeaMock).toHaveBeenCalledWith(ID);
      });
    });

    describe('and the component is double clicked', () => {
      it('should trigger an editIdea action', () => {
        const wrapper = shallow(<Idea {...requiredProps} />);

        wrapper.find('li').simulate('doubleClick');

        expect(editIdeaMock).toHaveBeenCalledWith(ID);
      });
    });
  });
});
