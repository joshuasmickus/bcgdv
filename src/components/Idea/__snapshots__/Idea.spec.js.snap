// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an Idea component when it is rendered should match the snapshot 1`] = `
<li
  className="idea some Id"
  onDoubleClick={[Function]}
>
  <h3>
    this title here
  </h3>
  <p>
    the body
  </p>
  <Button
    className="delete-button"
    onClick={[Function]}
  >
    <img
      alt="Delete"
      src="delete.png"
      title="Delete"
    />
  </Button>
</li>
`;
